File: plugin/src/main/java/com/stratio/cassandra/lucene/IndexWriter.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene;

import org.apache.cassandra.db.DecoratedKey;
import org.apache.cassandra.db.DeletionTime;
import org.apache.cassandra.db.RangeTombstone;
import org.apache.cassandra.db.rows.Row;
import org.apache.cassandra.index.transactions.IndexTransaction;
import org.apache.cassandra.utils.concurrent.OpOrder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * {@link Index.Indexer} for Lucene-based index.
 *
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
abstract class IndexWriter implements Index.Indexer {

    protected static final Logger logger = LoggerFactory.getLogger(IndexWriter.class);

    protected final IndexService service;
    protected final DecoratedKey key;
    protected final int nowInSec;
    protected final OpOrder.Group opGroup;
    protected final IndexTransaction.Type transactionType;

    /**
     * Abstract constructor.
     *
     * @param service the service to perform the indexing operation
     * @param key key of the partition being modified
     * @param nowInSec current time of the update operation
     * @param opGroup operation group spanning the update operation
     * @param transactionType what kind of update is being performed on the base data
     */
    IndexWriter(IndexService service,
                DecoratedKey key,
                int nowInSec,
                OpOrder.Group opGroup,
                IndexTransaction.Type transactionType) {
        this.service = service;
        this.key = key;
        this.nowInSec = nowInSec;
        this.opGroup = opGroup;
        this.transactionType = transactionType;
    }

    /** {@inheritDoc} */
    @Override
    public void begin() {
    }

    /** {@inheritDoc} */
    @Override
    public void partitionDelete(DeletionTime deletionTime) {
        logger.trace("Delete partition during {}: {}", transactionType, deletionTime);
        delete();
    }

    /** {@inheritDoc} */
    @Override
    public void rangeTombstone(RangeTombstone tombstone) {
        logger.trace("Range tombstone during {}: {}", transactionType, tombstone);
    }

    /** {@inheritDoc} */
    @Override
    public void insertRow(Row row) {
        logger.trace("Insert rows during {}: {}", transactionType, row);
        index(row);
    }

    /** {@inheritDoc} */
    @Override
    public void updateRow(Row oldRowData, Row newRowData) {
        logger.trace("Update row during {}: {} TO {}", transactionType, oldRowData, newRowData);
        index(newRowData);
    }

    /** {@inheritDoc} */
    @Override
    public void removeRow(Row row) {
        logger.trace("Remove row during {}: {}", transactionType, row);
        index(row);
    }

    /**
     * Deletes all the partition.
     */
    protected abstract void delete();

    /**
     * Indexes the specified partition's {@link Row}. It behaviours as an upsert and may involve read-before-write.
     *
     * @param row the row to be indexed.
     */
    protected abstract void index(Row row);
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/IndexWriter.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene

import com.stratio.cassandra.lucene.util.{Logging, Tracing}
import org.apache.cassandra.db._
import org.apache.cassandra.db.rows.{Row, RowIterator, UnfilteredRowIterators}
import org.apache.cassandra.index.Index.Indexer
import org.apache.cassandra.index.transactions.IndexTransaction
import org.apache.cassandra.index.transactions.IndexTransaction.Type.CLEANUP
import org.apache.cassandra.utils.concurrent.OpOrder

/** [[Indexer]] for Lucene-based index.
  *
  * @param service         the service to perform the indexing operation
  * @param key             key of the partition being modified
  * @param nowInSec        current time of the update operation
  * @param opGroup         operation group spanning the update operation
  * @param transactionType what kind of update is being performed on the base data
  * @author Andres de la Pena `adelapena@stratio.com`
  */
abstract class IndexWriter(
    service: IndexService,
    key: DecoratedKey,
    nowInSec: Int,
    opGroup: OpOrder.Group,
    transactionType: IndexTransaction.Type) extends Indexer with Logging with Tracing {

  val metadata = service.metadata
  val table = service.table

  /** @inheritdoc */
  override def begin() {
    logger.trace(s"Begin transaction $transactionType")
  }

  /** @inheritdoc */
  override def partitionDelete(deletionTime: DeletionTime) {
    logger.trace(s"Delete partition during $transactionType: $deletionTime")
    delete()
  }

  /** @inheritdoc */
  override def rangeTombstone(tombstone: RangeTombstone) {
    logger.trace(s"Range tombstone during $transactionType: $tombstone")
    delete(tombstone)
  }

  /** @inheritdoc */
  override def insertRow(row: Row): Unit = {
    logger.trace(s"Insert rows during $transactionType: $row")
    tryIndex(row)
  }

  /** @inheritdoc */
  override def updateRow(oldRowData: Row, newRowData: Row): Unit = {
    logger.trace(s"Update row during $transactionType: $oldRowData TO $newRowData")
    tryIndex(newRowData)
  }

  /** @inheritdoc */
  override def removeRow(row: Row): Unit = {
    logger.trace(s"Remove row during $transactionType: $row")
    tryIndex(row)
  }

  /** Deletes all the partition. */
  protected def delete()

  /** Deletes all the rows in the specified tombstone. */
  protected def delete(tombstone: RangeTombstone)

  /** Try indexing the row. If the row does not affect index it is not indexed */
  private[this] def tryIndex(row: Row): Unit = {
    if (service.doesAffectIndex(row)) {
      index(row)
    } else {
      tracer.trace("Lucene index skipping row")
    }
  }

  /** Indexes the specified row. It behaviours as an upsert and may involve read-before-write.
    *
    * @param row the row to be indexed.
    */
  protected def index(row: Row)

  /** Retrieves from the local storage the rows satisfying the specified read command.
    *
    * @param command a single partition read command
    * @return a row iterator
    */
  protected def read(command: SinglePartitionReadCommand): RowIterator = {
    val controller = command.executionController
    try {
      val unfilteredRows = command.queryMemtableAndDisk(table, controller)
      UnfilteredRowIterators.filter(unfilteredRows, nowInSec)
    } finally controller.close()
  }

  /** @inheritdoc */
  override final def finish() {

    // Skip on cleanups
    if (transactionType == CLEANUP) return

    // Finish with mutual exclusion on partition
    service.readBeforeWriteLocker.run(key, () => commit())
  }

  /** Commits all pending writes */
  protected def commit()

}